src/modules/roc_core/target_libunwind/roc_core/backtrace.cpp

```diff
@@ -0,0 +1,188 @@
+    /* Moving to previously called frames & going through each of them.
+     * unw_step() is signal safe.
+     * If there is atleast one entry in backtrace
+     */
+    if (unw_step(&cursor) > 0) {
+        return true;
+    }
+    return false;
+}
+
+/* Print function name, offset, instruction pointer address
+ * This must be signal safe.
+ */
+void backtrace_symbols() {
+    /* To store the snapshot of the CPU registers.
+     */
+    unw_context_t context;
+
+    /* To get snapshot of the CPU register.
+     * unw_getcontext() is signal safe.
+     */
+    unw_getcontext(&context);
+
+    /* To point to the current frame in the call stack
+     */
+    unw_cursor_t cursor;
+
+    /* To point to current frame.
+     * unw_init_local() is signal safe.
+     */
+    unw_init_local(&cursor, &context);
+
+    /* Moving to previously called frames & going through each of them.
+     * unw_step() is signal safe.
+     * buffer = final output that will be written to stderr.
+     */
+    uint32_t index = 0;
+    while (unw_step(&cursor) > 0) {
+        index++;
+        /* Get value stored in instruction register
+         * unw_get_reg() is signal safe.
+         */
+        unw_word_t ip;
+        unw_get_reg(&cursor, UNW_REG_IP, &ip);
+        /* Get function name & offset
+         * unw_get_proc_name() is signal safe.
+         * 'status' checks if unw_get_proc_name() is successful or not.
+         * 'status = 0', successfully executed the function.
+         */
+        char function_name[MaxLenFunctionName];
+        function_name[0] = '\0';
+        unw_word_t offset;
+        int32_t status =
+            unw_get_proc_name(&cursor, function_name, sizeof(function_name), &offset);
+        /* Printing => index, function name, offset, ip
+         * using write().
+         * write() & strcat() are signal safe.
+         */
+        if (status < 0) {
+            offset = 0;
+        }
+        char number[MaxDigits + 1];
+        uint32_to_string(index, 10, number);
+
+        char buffer[MaxLen] = "#";
+        size_t current_buffer_size = 1;
+        safe_strcat(buffer, current_buffer_size, number);
+        safe_strcat(buffer, current_buffer_size, ": (");
+        safe_strcat(buffer, current_buffer_size, function_name);
+        safe_strcat(buffer, current_buffer_size, "+0x");
+
+        uint32_to_string((uint32_t)offset, 16, number);
+        safe_strcat(buffer, current_buffer_size, number);
+        safe_strcat(buffer, current_buffer_size, ") [");
+
+        uint32_to_string((uint32_t)ip, 16, number);
+        safe_strcat(buffer, current_buffer_size, number);
+        safe_strcat(buffer, current_buffer_size, "]\n");
+
+        print_emergency_message(buffer);
+    }
+}
+} // namespace
+
+void print_backtrace() {
+    if (!is_backtrace_available()) {
+        fprintf(stderr, "No backtrace available\n");
+    } else {
+        fprintf(stderr, "Backtrace:\n");
+        backtrace_symbols();
+    }
+}
+
+void print_backtrace_emergency() {
```

Comment:
Did you test printing backtrace from a signal handler, e.g. on a crash?

It does not work for me. The backtrace does not print the pre-signal stack for some reason.

```
ERROR: caught SIGSEGV

#1: (_ZN3roc4core12_GLOBAL__N_114signal_handlerEiP9siginfo_tPv+0x6d) [6662a8fd]
#2: (sigwaitinfo+0x8) [e3de4e21]
#3: (+0x0) [e3d766d9]
```

Reply:
I think I figured it out. Is it because I'm always travelling through the stack and when a crash occurs, `unw_get_reg(&cursor, UNW_REG_IP, &ip)` will fail and just quit? If that is what's happening, then I should handle the case where it fails and print the rest of backtrace.

Reply:
Quick googling showed that we should not use `unw_getcontext()` for signal handler; instead, it seems that we should use `ucontext` passed to the handler.

I've tried this:

* added a `void* context` argument to `print_backtrace_emergency()`

* forwarded it from `signal_handler()` in `crash.cpp`:

      void signal_handler(int sig, siginfo_t* si, void* context) {
          ...
          print_backtrace_emergency(context);

* removed the local `context` declaration and `unw_getcontext()` call and used the provided pointer instead

And after this I saw the full backtrace.

Could you please check the docs that this is the recommended/supported use-case?

Note the we still need to initialize a local context using `unw_getcontext()` when the user calls `print_backtrace()` (regular call) instead of `print_backtrace_emergency()` (call from signal handler).

Reply:
I didn't find any doc using `ucontext_t` with `unw_context`.
In the unw_getcontext() [doc](https://www.nongnu.org/libunwind/man/unw_getcontext(3).html), it is given that unw_getcontext() is safe to be used from a signal handler. But passing the context as an argument to `print_backtrace_emergency()` does make sense.

Reply:
```
grep -E 'On.*ucontext_t' . -R
./libunwind-x86_64.h:118:/* On x86_64, we can directly use ucontext_t as the unwind context.  */
./libunwind-arm.h:253:/* On ARM, we define our own unw_tdep_context instead of using ucontext_t.
./libunwind-hppa.h:103:/* On PA-RISC, we can directly use ucontext_t as the unwind context.  */
./libunwind-ppc64.h:245:/* On ppc64, we can directly use ucontext_t as the unwind context.  */
./libunwind-ia64.h:150:/* On IA-64, we can directly use ucontext_t as the unwind context.  */
./libunwind-x86.h:164:/* On x86, we can directly use ucontext_t as the unwind context.  */
./libunwind-aarch64.h:172:/* On AArch64, we can directly use ucontext_t as the unwind context.  */
./libunwind-mips.h:130:/* On x86, we can directly use ucontext_t as the unwind context.  FIXME for
./libunwind-ppc32.h:181:/* On ppc, we can directly use ucontext_t as the unwind context.  */
```

Well, on most platforms ucontext_t can be used with libunwind, but still on some platforms (ARM) it can not.

Reply:
Hmmm.

I've checked out your branch once more and inserted a crash in a few other places and now I'm seeing the full backtrace without passing context from the signal handler.

Very strange, but I can't reproduce the issue anymore. I'll make a few more experiments, but if I will not reproduce this, and if it also works for you, let's keep it as is (without passing ucontext_t).

Reply:
Huh.

The backtrace is full on glibc but is truncated on alpine/musl.

I can reproduce it by inserting a crash somewhere (e.g. division by zero) and running tests in docker:

```
docker run --rm -t --cap-add SYS_PTRACE -u 1000 -v `pwd`:/opt/roc \
  -w /opt/roc rocproject/travis-alpine \
    scons -Q --enable-werror --build-3rdparty=openfec test
```

Reply:
Silly me! That's because I did not provide `--enable-debug` flag when building on alpine. I added it and it worked.

So, sorry for the confusion. The current libunwind version is good.

Except that the formatting is broken, I had to fix it:

```
@@ -33,7 +33,7 @@ void safe_strcat(char* buffer, size_t& buffer_size, const char* str) {
     if (len > 0) {
         memcpy(buffer + buffer_size, str, len);
         buffer_size += len;
-        buffer[MaxLen - 1] = '\0';
+        buffer[buffer_size] = '\0';
     }
 }
 
@@ -162,6 +162,7 @@ void backtrace_symbols() {
 
         uint32_to_string((uint32_t)offset, 16, number);
         safe_strcat(buffer, current_buffer_size, number);
+        safe_strcat(buffer, current_buffer_size, "\n");
 
         print_emergency_message(buffer);
     }
```